internal/pkg/service/stream/storage/node/coordinator/filerotation/operator.go

```diff
@@ -349,17 +338,25 @@ func (o *operator) closeFile(ctx context.Context, file *fileData) {
 	o.logger.Info(ctx, "closing file")
 
 	// Wait until all file slices are uploaded
-	err := o.waitForSlicesUpload(ctx, file.FileKey)
-	if err != nil {
+	if err := o.waitForSlicesUpload(ctx, file.FileKey); err != nil {
 		err = errors.PrefixError(err, "error when waiting for file slices upload")
 		o.logger.Error(ctx, err.Error())
 		return
 	}
 
-	// Find if file is empty
-	o.lock.RLock()
-	fileEmpty := !o.fileNotEmpty[file.FileKey]
-	o.lock.RUnlock()
+	// Make sure the statistics cache is up-to-date
+	if err := o.statisticsCache.WaitForRevision(ctx, file.ModRevision); err != nil {
+		err = errors.PrefixError(err, "error when waiting for statistics cache revision")
+		o.logger.Error(ctx, err.Error())
+		return
+	}
+
+	// Get file statistics
+	stats, err := o.statisticsCache.FileStats(ctx, file.FileKey)
+	if err != nil {
+		o.logger.Errorf(ctx, "cannot get file statistics: %s", err)
+		return
+	}
```

Comment:
I removed `fileEmpty` detection mechanism, and replaced it with statistics cache.

The `slicerotation` was using statistics cache to detect `isEmpty`, `filerotation` mirror - it should be same.

-----------------

To make the use of stats bulletproof, I added the `WaitForRevision` method.